stac_pydantic/api/utils/link_factory.py

```diff
@@ -0,0 +1,94 @@
+import inspect
+from dataclasses import dataclass
+from typing import List
+from urllib.parse import urljoin
+
+from ...links import Link, Links, Relations
+from ...shared import MimeTypes
+
+
+@dataclass
+class BaseLinks:
+    """Create inferred links common to collections and items."""
+
+    base_url: str
+    collection_id: str
+
+    def root(self) -> Link:
+        """Return the catalog root."""
+        return Link(
+            rel=Relations.root, type=MimeTypes.json, href=urljoin(self.base_url, "/")
+        )
+
+    def create_links(self) -> Links:
+        """Create inferred links"""
+        links = []
+        methods = inspect.getmembers(self, predicate=inspect.ismethod)
+        for method in methods:
+            func = method[-1]
+            ret_type = inspect.signature(func).return_annotation
+            # TODO: enforce full signature
+            if inspect.isclass(ret_type):
+                if issubclass(ret_type, Link):
+                    links.append(func())
+        return Links.parse_obj(links)
```

Comment:
Is this the simplest way to achieve this? It seems that you're trying to deduce all the link methods by introspecting the class. I'd personally just go for a class attribute like `_link_methods` that's a list of strings defining functions that create links. So for the `CollectionLinks` class, you'd have 

```py
@dataclass
class CollectionLinks(BaseLinks):
    """Create inferred links specific to collections."""

    _link_methods = ['parent', 'item']
```

and then `create_links` could be much simpler, something like:

```py
    def create_links(self) -> Links:
        """Create inferred links"""
        links = []
        for link_method in self._link_methods:
            links.append(getattr(self, link_method)())
```

It would require maintaining the `_link_methods` attribute but it would be less "magic"

Reply:
I think introspection is fine here, but I agree with you it's "magical" and is harder to understand if someone looks under the hood at how links are created.  `_link_methods` is a good approach too, I have to think more on this.

Reply:
I'm undecided on both of them but I hate `_link_methods` less, so let's go with that for now.  I'm not convinced this is the best way to create links, its just how arturo-stac-api does it.